service/matching/matchingEngine.go

```diff
@@ -321,6 +336,59 @@ pollLoop:
 	}
 }
 
+// QueryWorkflow creates a DecisionTask with query data, send it through sync match channel, wait for that DecisionTask
+// to be processed by worker, and then return the query result.
+func (e *matchingEngineImpl) QueryWorkflow(ctx context.Context, queryRequest *m.QueryWorkflowRequest) (*workflow.QueryWorkflowResponse, error) {
+	domainID := queryRequest.GetDomainUUID()
+	taskListName := *queryRequest.TaskList.Name
+	taskList := newTaskListID(domainID, taskListName, persistence.TaskListTypeDecision)
+	tlMgr, err := e.getTaskListManager(taskList)
+	if err != nil {
+		return nil, err
+	}
+	queryTask := &queryTaskInfo{
+		queryRequest: queryRequest,
+		taskID:       uuid.New(),
+	}
+	err = tlMgr.SyncMatchQueryTask(ctx, queryTask)
+	if err != nil {
+		return nil, err
+	}
+
+	queryResultCh := make(chan *workflow.RespondQueryTaskCompletedRequest)
+	e.queryMapLock.Lock()
+	e.queryTaskMap[queryTask.taskID] = queryResultCh
+	e.queryMapLock.Unlock()
+	defer func() {
+		e.queryMapLock.Lock()
+		delete(e.queryTaskMap, queryTask.taskID)
+		e.queryMapLock.Unlock()
+	}()
+
+	select {
+	case result := <-queryResultCh:
```

Comment:
if I understand right, the queryTask doesn't have a decision task timeout and this method relies on the caller to set a reasonable timeout value for the context to bail out ? if this is true, we should also include a timeout in the select{} in IMO

Reply:
the context here is from frontend, which create new context with default 1m timeout and also put the thrift call's context as parent context. so whichever timeout first will be the timeout here.